File: programmers/tired.java
import java.util.Arrays;

class Solution {
    int ans = -1;
    int arr[][];
    boolean check[];
    
    void dfs(int hp, int cnt){
        if(cnt > ans) ans = cnt;
        
        for(int i = 0; i < arr.length; i++){
            if(check[i] || arr[i][0] > hp) continue;
            check[i] = true;
            dfs(hp - arr[i][1], cnt+1);
            check[i] = false;
        }
    }
    
    public int solution(int k, int[][] dungeons) {
        arr = dungeons;
        check = new boolean[8+1];
        Arrays.fill(check, false);

        for(int i = 0; i < dungeons.length; i++){
            if(dungeons[i][0] > k) continue;
            check[i] = true;
            dfs(k - dungeons[i][1], 1);
            check[i] = false;
        }

        return ans;
    }
}

File: programmers/tired.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool comp(vector<int> a, vector<int> b){
    if(a[0] == b[0])
        return a[1] >= b[1];
    return a[0] > b[0];
}

int arr[1000+1];
int ans;

void dfs(vector<vector<int>> dungeons, vector<bool> check, int hp, int cnt){
    if(cnt > ans) ans = cnt;
    
    for(int i = 0; i < dungeons.size(); i++){
        if(check[i] || hp < dungeons[i][0]) continue;
        check[i] = true;
        dfs(dungeons, check, hp - dungeons[i][1], cnt+1);
        check[i] = false;
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    sort(dungeons.begin(), dungeons.end(), comp);
    vector<bool> check(8+1, false);
    
    for(int i = 0; i < dungeons.size(); i++){
        if(dungeons[i][0] > k) continue;
        check[i] = true;
        dfs(dungeons, check, k - dungeons[i][1], 1);
        check[i] = false;
    }
    return ans;
}
